main.go

```diff
@@ -269,11 +279,86 @@ func (fp *Firehose) RunLeader(ctx context.Context) {
 	}
 }
 
+func (fp *Firehose) IsLeader(ctx context.Context) (*bool, error) {
+	leader := true
+	nodeId := fp.storageBackend.GetNodeId(ctx)
+	deploymentId := fp.storageBackend.GetDeploymentId(ctx)
+	if nodeId == "" || deploymentId == "" {
+		panic("nodeId or deploymentId can not be determined")
+	}
+	leaderWriteErr := fp.storageBackend.WriteLeaderFile(ctx, fp.processSettings.MetadataBucket, nodeId, deploymentId)
+
+	_, ok := leaderWriteErr.(shared.LeaderFileExists)
+	if ok {
+		r, err := fp.storageBackend.CreateReader(ctx, fp.processSettings.MetadataBucket, "leader.json")
+		if err != nil {
+			fmt.Println(err)
+			return nil, err
+		}
+		data, err := ioutil.ReadAll(r)
+		if err != nil {
+			return nil, err
+		}
+		var result map[string]interface{}
+		err = json.Unmarshal(data, &result)
+		if err != nil {
+			return nil, err
+		}
+
+		if result["deploymentId"] != deploymentId {
+			return nil, fmt.Errorf("deploymentId does not match current leader")
+		}
+		if result["nodeId"] == nodeId && result["deploymentId"] == deploymentId {
+			return &leader, nil
+		}
+
+		leader = false
+		return &leader, nil
+	} else if leaderWriteErr == nil {
+		return &leader, nil
+	} else {
+		return nil, leaderWriteErr
+	}
+}
+
 // RunFirehose starts a Firehose running in leader of follower mode
-func (f *Firehose) RunFirehose() {
+func (fp *Firehose) RunFirehose(ctx context.Context) {
 	// 1. on start up determine if leader or followerBackend
-	// 2. if leader call RunLeader
-	// 3. else follower call RunFollower
+	globalTimeout := fp.processSettings.AcquireRoleTimeout
+	timeCheckingLeader := 0
+	var leader *bool
+outer:
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			// poll for valid isLeader every 2 seconds
+			interval := 2
+			<-time.After(time.Second * time.Duration(interval))
+			var err error
+			leader, err = fp.IsLeader(ctx)
+			if err != nil {
+				timeCheckingLeader = timeCheckingLeader + interval
+				if timeCheckingLeader >= globalTimeout {
+					panic("failed to read leader file or deploymentId did not match")
+				}
+				// retry if error reading leader file
+				continue
+			}
+			break outer
+		}
+	}
+	if *leader {
+		// 2. if leader call RunLeader
+		go fp.RunLeader(ctx)
```

Comment:
shouldn't this be sync call (same as follower)?

Reply:
yeah I think that makes sense now that I changed it to be an infinite loop in the previous